Tidy up instance time handling in SVGSMILElement

Ahmad-S792 opened the pull request:
#### 1465a7975860d39d4b6e1a2f8d2dd4e6a77a3e92
<pre>
Tidy up instance time handling in SVGSMILElement

<a href="https://bugs.webkit.org/show_bug.cgi?id=275296">https://bugs.webkit.org/show_bug.cgi?id=275296</a>

Reviewed by Darin Adler.

Merge: <a href="https://github.com/chromium/chromium/commit/901d1e45d6a0a49407a7e8936f6edf9b6124c260">https://github.com/chromium/chromium/commit/901d1e45d6a0a49407a7e8936f6edf9b6124c260</a>

Merge the addBeginTime and addEndTime methods into a single method
addInstanceTime (that takes a BeginOrEnd argument.) Migrate callers to
the new method, getting rid of SVGSMILElement::handleConditionEvent in
the process.

Also inline some trivial functions in SVGAnimationElement.

* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::getSimpleDuration const):
(WebCore::SVGAnimationElement::beginElementAt):
(WebCore::SVGAnimationElement::endElementAt):
(WebCore::SVGAnimationElement::beginElement): Deleted.
(WebCore::SVGAnimationElement::endElement): Deleted.
* Source/WebCore/svg/SVGAnimationElement.h:
(WebCore::SVGAnimationElement::beginElement):
(WebCore::SVGAnimationElement::endElement):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::ConditionEventListener::handleEvent):
(WebCore::SVGSMILElement::addInstanceTime):
(WebCore::SVGSMILElement::createInstanceTimesFromSyncbase):
(WebCore::SVGSMILElement::removeTimeDependent):
(WebCore::SVGSMILElement::beginByLinkActivation):
(WebCore::SVGSMILElement::addBeginTime): Deleted.
(WebCore::SVGSMILElement::addEndTime): Deleted.
(WebCore::SVGSMILElement::handleConditionEvent): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.h:

Canonical link: <a href="https://commits.webkit.org/279971@main">https://commits.webkit.org/279971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55a14aeac70bcb36689231445845e4bbfb6b9eb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55039 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7650 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57339 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44562 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3937 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57134 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47693 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25686 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5038 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3915 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59911 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30314 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5417 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51993 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31441 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51447 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12108 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32468 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31232 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -119,8 +119,9 @@ bool ConditionEventListener::operator==(const EventListener& listener) const
-        animation->handleConditionEvent(m_condition);
+    if (!m_animation)
+        return;
+    m_animation->addInstanceTime(m_condition->m_beginOrEnd, m_animation->elapsed() + m_condition->m_offset);
```
Not good to remove the object lifetime change where we keep the animation in a local `RefPtr`.

Ahmad-S792 replied:
Done - reintroduced it in https://github.com/WebKit/WebKit/commit/f397d57751ebb138f81c2c693fdbd4bab66218b7.

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -751,20 +752,18 @@ SMILTime SVGSMILElement::simpleDuration() const
+    SMILTime elapsed = this->elapsed();
+    if (elapsed.isUnresolved())
+        return;
+    Vector<SMILTimeWithOrigin>& list = beginOrEnd == Begin ? m_beginTimes : m_endTimes;
```
I’d suggest `auto&` here.

Ahmad-S792 replied:
Done - in https://github.com/WebKit/WebKit/commit/f397d57751ebb138f81c2c693fdbd4bab66218b7.

darinadler reviewed (changes requested):
Not sure this is a significant improvement. Mostly a set of arbitrary changes, only very slightly tidier. At the very least, we need to add back the `RefPtr`.

darinadler commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -120,7 +120,7 @@ bool ConditionEventListener::operator==(const EventListener& listener) const
 {
     if (RefPtr animation = m_animation.get())
-        animation->handleConditionEvent(m_condition);
+        m_animation->addInstanceTime(m_condition->m_beginOrEnd, m_animation->elapsed() + m_condition->m_offset);
```
should be `animation` here not `m_animation`

Ahmad-S792 replied:
Ops! Sorry! Missed it. Done in [55a14ae](https://github.com/WebKit/WebKit/commit/55a14aeac70bcb36689231445845e4bbfb6b9eb7).

Commit:
Tidy up instance time handling in SVGSMILElement

https://bugs.webkit.org/show_bug.cgi?id=275296

Reviewed by Darin Adler.

Merge: https://github.com/chromium/chromium/commit/901d1e45d6a0a49407a7e8936f6edf9b6124c260

Merge the addBeginTime and addEndTime methods into a single method
addInstanceTime (that takes a BeginOrEnd argument.) Migrate callers to
the new method, getting rid of SVGSMILElement::handleConditionEvent in
the process.

Also inline some trivial functions in SVGAnimationElement.

* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::getSimpleDuration const):
(WebCore::SVGAnimationElement::beginElementAt):
(WebCore::SVGAnimationElement::endElementAt):
(WebCore::SVGAnimationElement::beginElement): Deleted.
(WebCore::SVGAnimationElement::endElement): Deleted.
* Source/WebCore/svg/SVGAnimationElement.h:
(WebCore::SVGAnimationElement::beginElement):
(WebCore::SVGAnimationElement::endElement):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::ConditionEventListener::handleEvent):
(WebCore::SVGSMILElement::addInstanceTime):
(WebCore::SVGSMILElement::createInstanceTimesFromSyncbase):
(WebCore::SVGSMILElement::removeTimeDependent):
(WebCore::SVGSMILElement::beginByLinkActivation):
(WebCore::SVGSMILElement::addBeginTime): Deleted.
(WebCore::SVGSMILElement::addEndTime): Deleted.
(WebCore::SVGSMILElement::handleConditionEvent): Deleted.
* Source/WebCore/svg/animation/SVGSMILElement.h:

Canonical link: https://commits.webkit.org/279971@main
